### PR TITLE
PERF: Use a C Union in maybe_convert_numeric

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

```
       before           after         ratio
     [22de62ca]       [3a53d05b]
                      <maybe-convert-less-mem>
-            743M             673M     0.91  inference.MaybeConvertNumeric.peakmem_convert('float64')
-            986M             836M     0.85  inference.MaybeConvertNumeric.peakmem_convert('int64')
-            986M             836M     0.85  inference.MaybeConvertNumeric.peakmem_convert('uint64')
-        685±10ms         500±30ms     0.73  inference.MaybeConvertNumeric.time_convert('bool')
-       2.20±0.1s       1.51±0.07s     0.69  inference.MaybeConvertNumeric.time_convert('uint64')
-            428M             278M     0.65  inference.MaybeConvertNumeric.peakmem_convert('bool')
-      2.07±0.08s       1.23±0.06s     0.60  inference.MaybeConvertNumeric.time_convert('int64')
```

### CODE REVIEW

File: pandas/core/tools/numeric.py [hunk 1]
```diff
@@ -205,6 +205,9 @@ def to_numeric(
                 coerce_numeric=coerce_numeric,
                 convert_to_masked_nullable=use_nullable_dtypes,
             )
+            #print(f"Values is {values}")
+            #print(f"Values dtype is {values.dtype}")
+            #print(f"Mask is {new_mask}")
```
Comment (jbrockmendel) [inline pandas/core/tools/numeric.py:210]: pls remove

**Comment by jbrockmendel** [comment]
ballpark how much improvement?

File: pandas/core/tools/numeric.py (hunk 1 above)
Comment (lithomas1) [reply pandas/core/tools/numeric.py:210]: thx, removed.

**Comment by lithomas1** [comment]
> ballpark how much improvement?

Updated the OP.

Note, though, that I've turned off the fastpath for ints in my benchmarks.

I would like @WillAyd 's thoughts on this, though to make sure I'm not doing something sketchy with the union or the casting in the Cython code, though.

File: pandas/_libs/lib.pyx [hunk 2]
```diff
@@ -2124,6 +2128,56 @@ cpdef bint is_interval_array(ndarray values):
...
+    bool_t = 0
+    null_t = 1
+    uint_t = 2
+    int_t = 3
+    float_t = 4
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef void convert_union_arr_to_ndarray(NonComplexNumeric* arr,
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:2154]: Not super familiar with this code by is there no way for us to pass a fused-type specialized memoryview or ndarray?

File: pandas/_libs/lib.pyx [hunk 3]
```diff
@@ -2189,23 +2243,25 @@ def maybe_convert_numeric(
...
+        # Keeps track of what element type is stored in each index
+        # Also tracks nulls for masked arrays
+        # (We'll fix mask if convert_to_masked_nullable=True)
+        # See convert_union_arr_to_ndarray
+        ndarray[uint8_t, ndim=1] mask = cnp.PyArray_EMPTY(
             1, values.shape, cnp.NPY_UINT8, 0
         )
-        ndarray[uint8_t, ndim=1] mask = np.zeros(n, dtype="u1")
+        NonComplexNumeric *arr = <NonComplexNumeric *>PyMem_Malloc(
+            n * sizeof(NonComplexNumeric)
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:2261]: A union takes up the size of its largest element, so this doesn't really save anything over the previous approach

I'm also a bit wary of the PyMem_Malloc because it seems like there are quite a few code paths that would return with a memory leak

Comment (lithomas1) [reply pandas/_libs/lib.pyx:2261]: From my understanding, the previous code allocated separate arrays for ints, floats, bools, etc, and did something like this:

See complex num: update complexes[i]
See float: update floats[i] and complexes[i]
See int: update floats[i], complexes[i]
See bool: update ints[i], floats[i], complexes[i]

So by using 1 array, we would save quite a bit for ints and bools in terms of peak memory usage, but not really that much for floats.

When trying to optimize this, I originally tried to use an object array to store elements and astype, but it turns out that that resulted in a regression with floats ( both memory-wise and speed-wise), which I think is caused by the fact that PyObjects are big in terms of byte sizes.

Comment (WillAyd) [reply pandas/_libs/lib.pyx:2261]: Ah OK yea that is really inefficient. Nice find. And yea I wouldn't expect the PyObject case to necessarily be faster since you are going from an array of primitive types to an array of PyObject* that still point elsewhere in the heap for value retrieval

That said, I'm not sure about this resolution. I think you'd be better served refactoring this function and trying to dispatch to fused types when we know the required type

Comment (jbrockmendel) [reply pandas/_libs/lib.pyx:2261]: > I think you'd be better served refactoring this function and trying to dispatch to fused types when we know the required type

IIUC the suggestion is a bit like infer_dtype or maybe_convert_objects where we iterate until we see a non-numeric entry and then dispatch based on that entry's type.  if we can use that pattern for the numeric types too that'd be great.

Comment (lithomas1) [reply pandas/_libs/lib.pyx:2261]: If I'm not mistaken, ``maybe_convert_objects`` has the exact same pattern as ``maybe_convert_numeric``. 

The problem here is that unless we see a complex, we can't be sure of the return type until after 1 full pass, cause an element at the end could force an upcast.

The alternative to this method would be to iterate a second time through the array, and convert based on mask values set by the first pass, so we don't have to infer again but this may have a perf penalty (probably would result in better memory usage since we wouldn't have an intermediate array).

Comment (WillAyd) [reply pandas/_libs/lib.pyx:2261]: Can you not just astype to a complex dtype when/if you come across it?

**Comment by lithomas1** [comment]
Reworked the approach to upcast when necessary. 
There's still a lot of boilerplate, though.
(I'm hoping that we can refactor some of it to live on the ``Seen`` object so that it can be shared with ``maybe_convert_objects``)

Perf is slightly better than the old version.
(I suspect that the union array approach could be faster if we did a zero copy creation of the ndarray from the union array, but that probably makes it way too complex). 

```
       before           after         ratio
     [0cedcbf6]       [fd205e6a]
                      <maybe-convert-less-mem>
-        664±20ms          525±7ms     0.79  inference.MaybeConvertNumeric.time_convert('bool')
-            743M             583M     0.78  inference.MaybeConvertNumeric.peakmem_convert('float64')
-            986M             746M     0.76  inference.MaybeConvertNumeric.peakmem_convert('uint64')
-            986M             746M     0.76  inference.MaybeConvertNumeric.peakmem_convert('int64')
-      2.12±0.04s       1.36±0.01s     0.64  inference.MaybeConvertNumeric.time_convert('uint64')
-      1.95±0.04s       1.24±0.02s     0.64  inference.MaybeConvertNumeric.time_convert('int64')
-            428M             268M     0.63  inference.MaybeConvertNumeric.peakmem_convert('bool')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

File: pandas/_libs/lib.pyx [hunk 4]
```diff
@@ -1356,6 +1380,29 @@ cdef class Seen:
             self.datetime_ or self.datetimetz_ or self.nat_ or self.timedelta_
             or self.period_ or self.interval_ or self.numeric_ or self.object_
         )
+    cdef void get_curr_val(self, curr_seen_t *ret_val):
+        """
+        Gets the current seen val and casts it to ret_val's type.
+        Caller is responsible for making sure that the cast is safe
+        """
+        if self.curr_seen is CurrSeen.bool_t:
+            ret_val[0] = <curr_seen_t>self.curr_seen_val.bool_t
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:1389]: Instead of this is there a way to convert this into a class that can manage the state of the known type as well as the corresponding data? I think if you go with a class-based approach you can more naturally achieve the polymorphic behavior you are ultimately after, but right now its scattered across the implementation.

Comment (lithomas1) [reply pandas/_libs/lib.pyx:1394]: In the latest commit, I've moved handling of setting item values to the Seen object, so right now everything is completely encapsulated within the Seen class. Is this what you were thinking of?

File: pandas/_libs/lib.pyx [hunk 5]
```diff
@@ -1235,6 +1257,8 @@ cdef class Seen:
         bint datetimetz_      # seen_datetimetz
         bint period_          # seen_period
         bint interval_        # seen_interval
+        CurrSeen curr_seen    # Type of object we just saw
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:1260]: Since there are individual members already on the class used for type-introspection, this seems pretty duplicative. Are the existing members not usable for the goal here? If not, we probably new a redesigned class structure for `Seen` instead of just adding this

File: pandas/_libs/lib.pyx [hunk 6]
```diff
@@ -1351,6 +1375,52 @@ cdef class Seen:
...
             or self.period_ or self.interval_ or self.numeric_ or self.object_
         )
+    cdef void get_curr_val(self, curr_seen_t *ret_val):
+        """
+        Gets the current seen val and casts it to ret_val's type.
+        Caller is responsible for making sure that the cast is safe
+        """
+        if self.curr_seen is CurrSeen.null_t:
+            if curr_seen_t is float64_t or curr_seen_t is complex128_t:
+                ret_val[0] = <curr_seen_t>NaN
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:1385]: I really want to avoid casting like this - I think it will be really difficult to maintain this lack of type safety over time. It _might_ work better if the cast were contained within a class that also manages the type, but right now that is spread across a few objects

File: doc/source/whatsnew/v2.0.0.rst [hunk 7]
```diff
@@ -1155,7 +1155,7 @@ Performance improvements
 - Fixed a reference leak in :func:`read_hdf` (:issue:`37441`)
 - Fixed a memory leak in :meth:`DataFrame.to_json` and :meth:`Series.to_json` when serializing datetimes and timedeltas (:issue:`40443`)
 - Decreased memory usage in many :class:`DataFrameGroupBy` methods (:issue:`51090`)
--
+- Performance improvement in :func:`to_numeric` with object-datatype arrays (:issue:`51029`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v2.0.0.rst:1158]: datatype->dtype?

File: pandas/_libs/lib.pyx [hunk 8]
```diff
@@ -1210,6 +1210,28 @@ try:
...
+    uint_t = 2
+    int_t = 3
+    float_t = 4
+    complex_t = 5
+
+cdef union CurrSeenVal:
+    uint8_t bool_t
+    uint64_t uint_t
+    int64_t int_t
+    float64_t float_t
```
Comment (jbrockmendel) [inline pandas/_libs/lib.pyx:1225]: i dont think we use bool_t, uint_t, int_t, float_t elsewhere.  can we avoid these?

File: pandas/_libs/lib.pyx [hunk 9]
```diff
@@ -1351,6 +1375,52 @@ cdef class Seen:
             self.datetime_ or self.datetimetz_ or self.nat_ or self.timedelta_
             or self.period_ or self.interval_ or self.numeric_ or self.object_
         )
+    cdef void get_curr_val(self, curr_seen_t *ret_val):
```
Comment (jbrockmendel) [inline pandas/_libs/lib.pyx:1378]: missing newline?

File: pandas/tests/dtypes/test_inference.py [hunk 10]
```diff
@@ -607,6 +607,20 @@ def test_convert_infs(self):
...
+            np.array([1 + 2j, 2 + 3j, 3 + 4j, 4 + 5j], dtype="O"),
+            np.array([1 + 2j, 2 + 3j, None, 4 + 5j], dtype="O"),
+            np.array([1 + 2j, 2, 3.14, np.complex128(np.nan)], dtype="O"),
+            np.array([2, 1 + 2j, 3.14, np.complex128(np.nan)], dtype="O"),
+        ],
+    )
+    def test_convert_complexes(self, arr):
+        result, _ = lib.maybe_convert_numeric(arr, set())
+        expected = arr.astype(np.complex128)
+        tm.assert_numpy_array_equal(result, expected)
```
Comment (jbrockmendel) [inline pandas/tests/dtypes/test_inference.py:622]: is this a bugfix?  if so, can you either a) separate PR or b) comment (in GH) on where the relevant edits are

**Comment by jbrockmendel** [comment]
did you ever try out the two-pass approach?  i.e. do inference first, then allocate just one array?

**Comment by lithomas1** [comment]
> did you ever try out the two-pass approach? i.e. do inference first, then allocate just one array?

I forgot that ``maybe_convert_numeric`` also can process strings with the ``floatify`` function, so we would need to store the parsed values somewhere, or do a double parse (can be expensive if all strings, which I don't think is uncommon).

Right now, we use just 1 array, but astype it whenever we need to upcast. IMO, this is about the same complexity as the other approach, a lot of the diff just comes from me moving logic to the ``Seen`` class, so ``maybe_convert_objects`` will be easier to port to this impl.

File: pandas/_libs/lib.pyx (hunk 5 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:1260]: The existing members are for what is seen in the array as a whole, while this is just for the current element.

I will convert the existing members to the enum(probably going to rename it to ``SeenDtype``) when I start changing ``maybe_convert_objects`` (there are a couple extra ones, e.g. ``seen.datetimetz_`` that that function uses).

File: pandas/_libs/lib.pyx (hunk 6 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:1385]: Hm. I don't see a good way to do this without Tempita. Fused types don't work in cdef classes IIRC. 

Would raising an error if downcasting happens be OK?

(Not sure about allowing uint64 <-> int64 since that's slightly trickier, and is handled separately.)

File: pandas/_libs/lib.pyx [hunk 11]
```diff
@@ -1351,6 +1375,66 @@ cdef class Seen:
...
+        """
+        if self.curr_seen is CurrSeen.null_:
+            if curr_seen_t is float64_t or curr_seen_t is complex128_t:
+                ret_val[0] = <curr_seen_t>NaN
+        if self.curr_seen is CurrSeen.bool_:
+            # There's nothing to a bool can be downcast to, it is 1 byte
+            # so no checking
+            ret_val[0] = <curr_seen_t>self.curr_seen_val.bool_
+        elif self.curr_seen is CurrSeen.uint_:
+            if sizeof(ret_val[0]) < sizeof(uint64_t):
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:1394]: Does sizeof work with an expression? Feels a little off from C where its a compile-time operator, but not familiar with what Cython may be doing

File: pandas/_libs/lib.pyx [hunk 12]
```diff
@@ -1351,6 +1375,66 @@ cdef class Seen:
...
+                raise ValueError("Downcasting uint64 is not allowed")
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:1395]: Do we need to add `except` to the function signature? Want to make sure these don't get swallowed by the cdef

File: pandas/_libs/lib.pyx [hunk 13]
```diff
@@ -1351,6 +1375,66 @@ cdef class Seen:
...
+        elif self.curr_seen is CurrSeen.float_:
+            if sizeof(ret_val[0]) < sizeof(float64_t) \
+                    or curr_seen_t is int64_t or curr_seen_t is uint64_t:
+                raise ValueError("Downcasting float64 is not allowed")
+            ret_val[0] = <curr_seen_t>self.curr_seen_val.float_
+        elif self.curr_seen is CurrSeen.complex_:
+            if curr_seen_t is complex128_t:
+                # Need to gate behind specialization since complex can't safely
+                # cast to other types throwing a compile error
+                ret_val[0] = <curr_seen_t>self.curr_seen_val.complex_
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:1410]: Are the casts here required? I think the types should work naturally without casting?

File: pandas/_libs/lib.pyx [hunk 14]
```diff
@@ -1351,6 +1375,66 @@ cdef class Seen:
             self.datetime_ or self.datetimetz_ or self.nat_ or self.timedelta_
             or self.period_ or self.interval_ or self.numeric_ or self.object_
         )
+    cdef void get_curr_val(self, curr_seen_t *ret_val):
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:1378]: Does the getter here have to assign to ret_val? I think would be more natural to have the return type be `curr_seen_t` instead of `void`

File: pandas/_libs/lib.pyx [hunk 15]
```diff
@@ -1210,6 +1210,28 @@ try:
 except AttributeError:
     pass
 
+cdef enum CurrSeen:
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:1213]: Can you combine `CurrSeen` and `CurrSeenVal` into one struct? I think it would help readability to manage all within one object instead of separate

**Review by WillAyd — COMMENTED** [review]
This is definitely heading in the right direction. I think there's still a few things we can do to help the readability and reduce casting. 

This is one of those things I think is more natural in plain C than Cython...so hope we can find the right middle ground

File: pandas/_libs/lib.pyx (hunk 11 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:1394]: Yes, I checked the generated C code. It's turns into a C sizeof.

File: pandas/_libs/lib.pyx (hunk 12 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:1395]: I don't think so. I hit these exceptions when I was first writing this up :).

File: pandas/_libs/lib.pyx (hunk 13 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:1410]: Removed everything except for the cast from float to e.g. int (This can't happen, it would hit the exception, but Cython doesn't know that.).

File: pandas/_libs/lib.pyx (hunk 14 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:1378]: I use the pointer so that Cython can know which fused variant to dispatch to. I don't think it's possible/good to have a variable return type.

File: pandas/_libs/lib.pyx (hunk 11 above)
Comment (WillAyd) [reply pandas/_libs/lib.pyx:1394]: Ah OK I see what Cython is doing. Makes sense, but I think it would be less fragile if you just did `sizeof(curr_seen_t)`

File: pandas/_libs/lib.pyx (hunk 13 above)
Comment (WillAyd) [reply pandas/_libs/lib.pyx:1410]: I might be misreading it but did you push this up? Still seeing all the casts?

File: pandas/_libs/lib.pyx (hunk 14 above)
Comment (WillAyd) [reply pandas/_libs/lib.pyx:1378]: Cython suggests you should return a value:

https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values

Otherwise you will do a `PyErr_Occurred()` check after each invocation, which is expensive compared to a primitive value check

File: pandas/_libs/lib.pyx (hunk 13 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:1410]: Sorry, it should be up now.

File: pandas/_libs/lib.pyx (hunk 15 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:1215]: I just pushed something, so that NA value handling goes through the ``Seen`` object as well. 

The logic should be greatly simplified now since everything goes through ``Seen`` now.

Comment (WillAyd) [reply pandas/_libs/lib.pyx:1215]: Very nice - seems like a much better home for this

File: pandas/_libs/lib.pyx (hunk 14 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:1386]: Do I just have to add ``except -1`` to the signature and change the return type to int?

I'm a little confused by the Cython docs ( I assume Cython adds ``return -1`` or something after each exception in the generated code when I do this?).

Comment (WillAyd) [reply pandas/_libs/lib.pyx:1386]: Yea I think you have the right idea in mind. I would explicitly return either 0 or the value itself when all is well. I think Cython will automatically return -1 on exception and check in any caller

File: pandas/_libs/lib.pyx (hunk 6 above)
Comment (WillAyd) [reply pandas/_libs/lib.pyx:1396]: Really nice job getting rid of the casting. What happens if you remove it on this last one though? Not immediately obvious why this would still be necessary

File: pandas/_libs/lib.pyx [hunk 16]
```diff
@@ -1353,6 +1383,67 @@ cdef class Seen:
             self.datetime_ or self.datetimetz_ or self.nat_ or self.timedelta_
             or self.period_ or self.interval_ or self.numeric_ or self.object_
         )
+    cdef int get_curr_val(self, curr_seen_t *ret_val) except -1:
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:1386]: This function acts as a setter not a getter. Probably better named as `set_current_value`

File: pandas/_libs/lib.pyx [hunk 17]
```diff
@@ -1353,6 +1383,67 @@ cdef class Seen:
...
+        elif self.curr_seen is CurrSeen.complex_:
+            if curr_seen_t is complex128_t:
+                # Need to gate behind specialization since complex can't safely
+                # cast to other types throwing a compile error
+                ret_val[0] = self.curr_seen_val.complex_
+            else:
+                raise ValueError("Downcasting complex128 is not allowed")
+        return 0  # Success
+
+    cdef void set_curr_val(self, curr_seen_t val):
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:1423]: The semantics are a bit murky here since the function preceding this is also a setter. What is this supposed to do differently?

File: pandas/_libs/lib.pyx [hunk 18]
```diff
@@ -1299,12 +1323,18 @@ cdef class Seen:
...
                 and not self.coerce_numeric)
 
-    cdef saw_null(self):
+    cdef saw_null(self, seen_float=True):
         """
         Set flags indicating that a null value was encountered.
+
+        Parameters
+        ----------
+        seen_float: boolean, default True
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:1332]: The docstring here is a little unclear; the keyword is `seen_float` but this makes it seem like it should be called `seen_nan`, though that's also a bit confusion since the entire function is `saw_null`.

File: pandas/_libs/lib.pyx (hunk 17 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:1423]: This one sets the value on the Seen object, so it is actually a setter.

File: pandas/_libs/lib.pyx (hunk 18 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:1332]: ``seen_float`` is there to not break ``maybe_convert_objects``. For some reason, ``seen_null`` today means ``seen_nan``(floating NaN)

In the future, I plan on getting rid of this parameter. (It's probably better to go through ``seen.set_curr_val`` for floating NaNs)

File: pandas/_libs/lib.pyx (hunk 16 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:1386]: Well, technically it gets the value from the Seen object. Assigning to the pointer is awkward, but it's better than having a dummy parameter just for the sake of dispatching correctly.

Maybe there's a better style for doing this though?

File: pandas/_libs/lib.pyx (hunk 6 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx]: Should be able to get rid of that one too. thanks.

File: pandas/_libs/lib.pyx (hunk 16 above)
Comment (WillAyd) [reply pandas/_libs/lib.pyx:1386]: I think you can still make this a getter if you remove the pointer reference and just do `complexes[i] = self.get_curr_val()`

Comment (lithomas1) [reply pandas/_libs/lib.pyx:1397]: Sorry, I don't think I was being clear/thinking clearly when I typed that. I'll write a more in-depth explanation below.

To recap, the current approach requires a pointer since I am currently using a fused type(to reduce code duplication).
In Cython, fused functions need to have a fused type as a parameter, so that Cython knows which variant to dispatch to. So, our function signature has to look something like this
```
cdef int get_curr_val(self, curr_seen_t *ret_val) except -1: # Current
cdef curr_seen_t get_curr_val(self, curr_seen_t *ret_val): # If we were to return curr_seen_t, we lose error handling here
```

While thinking about what you said, though, maybe there is a third approach using void pointers.
Then, our signature would look like this
```
cdef void* get_curr_val(self, CurrSeen ret_type)
```
Here, by using void* instead of a fused function, we avoid having to pass a curr_seen_t pointer/regular type parameter into the function.

This function would then get the currently seen value, and then cast it to the return type (while checking that we are doing a correct cast like before), and return a pointer to it in the form of a void pointer.

I think we would be able to keep the error checking here (by using NULL as the error value).
Then, we would do e.g. 
```
complexes[i] = ((complex *)seen.get_curr_val())[0]
```
This is a lot more uglier than ``seen.get_curr_val(&complexes[i])``, but maybe it is better C style (it also gets rid of the fused functions) ?

Comment (WillAyd) [reply pandas/_libs/lib.pyx:1397]: I'm not sure I follow why you would lose error handling when returning `curr_seen_t`. If you use `except *` it should add a `PyErr_Occurred()` check after each invocation. That adds a performance penalty compared to a more targeted approach, but not sure how much it would be in the grand scheme of this refactor

Comment (lithomas1) [reply pandas/_libs/lib.pyx:1397]: Sorry, I forgot about that case (would still like to avoid, though).

I think my point about the dummy parameters still stands, though.

Comment (WillAyd) [reply pandas/_libs/lib.pyx:1397]: I think it would be simpler if you just created `get_{int|uint|float|complex}_value` functions. I know it seems  basic compared to fused types but injecting an arbitrary parameter to appease the templating requirements really hurts the readability.

Comment (lithomas1) [reply pandas/_libs/lib.pyx]: I've updated the code to not use the fused function now. Performance seems to be unaffected.

File: pandas/_libs/lib.pyx [hunk 19]
```diff
@@ -1365,6 +1395,89 @@ cdef class Seen:
...
+        elif self.curr_seen is CurrSeen.bool_:
+            return self.curr_seen_val.bool_
+        elif self.curr_seen is CurrSeen.float_:
+            return self.curr_seen_val.float_
+        else:
+            # complex cannot be cast to float
+            raise ValueError("Cannot convert complex to float64")
+
+    cdef complex128_t get_complex_val(self):
+        if self.curr_seen is CurrSeen.null_:
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:1442]: Shouldn't these just have one condition to match the function name? Was expecting `get_complex_val` to just do:

```
if self.curr_seen is CurrSeen.complex_:
    return self.curr_seen_val.complex_
else:
    raise ValueError(...)
```

Comment (lithomas1) [reply pandas/_libs/lib.pyx:1442]: No, since we can upcast nan/bool/int/uint/float to complex128.

Comment (lithomas1) [reply pandas/_libs/lib.pyx:1442]: @WillAyd 
Maybe renaming these to ``as_complex``, ``as_float``, etc. would make more sense?

(I should've been more specific in my last comment, but the case that these would be triggered by is when e.g. you have [1.0, 2, 3]. When the curr_val becomes 2, you'll want to get the value as a float(even though curr_val is an int), since the array can only be inferred as a float/complex array at this point)

Comment (WillAyd) [reply pandas/_libs/lib.pyx:1442]: Ah ok that is helpful thanks. Yea I think your suggestion makes sense. Do we also need each as_ function to have its own set of if conditions or can they be structured in a way that they can generically be placed in any implementation, just leaving it to the function to short circuit appropriately? 

Reason I ask is that the current structure generates a hierarchy of "available types" but in a way I don't think is very obvious. Probably better to trade some precision for readability/maintenance

Comment (lithomas1) [reply pandas/_libs/lib.pyx:1442]: > Ah ok that is helpful thanks. Yea I think your suggestion makes sense. Do we also need each as_ function to have its own set of if conditions or can they be structured in a way that they can generically be placed in any implementation, just leaving it to the function to short circuit appropriately?
> 
> Reason I ask is that the current structure generates a hierarchy of "available types" but in a way I don't think is very obvious. Probably better to trade some precision for readability/maintenance

Yeah, I'm not too much too much a fan of this as well (unfortunately I don't think the current behavior can be changed). 
I'm also not thinking of a way to simplify it on top of my head, since use of the union requires the series of if statements. 

This hopefully shouldn't cause too big of an issue in the future, since
1) The numpy dtypes are unlikely to change, so this code shouldn't change much
2) Our test coverage should be good enough for this.

I've also embedded some comments in each function listing the rejected types for each function.

Given this, would you be fine merging this in its current state?

**Comment by WillAyd** [comment]
I'll take another pass at it this weekend. @jbrockmendel care to take a look as well?

File: pandas/_libs/lib.pyx [hunk 20]
```diff
@@ -2227,123 +2371,146 @@ def maybe_convert_numeric(
...
         elif util.is_integer_object(val):
-            floats[i] = complexes[i] = val
-
             val = int(val)
             seen.saw_int(val)
 
             if val >= 0:
                 if val <= oUINT64_MAX:
-                    uints[i] = val
+                    seen.set_curr_val(<uint64_t>val)
```
Comment (WillAyd) [inline pandas/_libs/lib.pyx:2394]: Are these casts necessary? I thought we got rid of them before - maybe a mistake in merge that brought them back? Generally should not have a need for casting - only weakens the type system further

**Review by WillAyd — COMMENTED** [review]
I am a little concerned about the structure of the logic. Its hard to follow but also a complex change, so I'm not sure where the right balance is. 

If anyone else from @pandas-dev/pandas-core could take a look would be helpful

File: pandas/_libs/lib.pyx (hunk 20 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:2394]: Thanks for the review.
Casting should be required here. 
val is a Python object, so Cython wouldn't know how to implicitly cast here.
(We also do special things like casting bool -> uint8, and the logic to decide to cast to int64/uint64 you're referring to here).

File: pandas/_libs/lib.pyx (hunk 18 above)
Comment (jbrockmendel) [reply pandas/_libs/lib.pyx:1343]: agreed it would be nice to make the naming/docs unambiguous.  I'm a big fan of breaking things like this off into precursor PRs

File: pandas/_libs/lib.pyx [hunk 21]
```diff
@@ -1223,6 +1223,28 @@ try:
 except AttributeError:
     pass
 
+cdef enum CurrSeen:
+    bool_ = 0
+    null_ = 1
```
Comment (jbrockmendel) [inline pandas/_libs/lib.pyx:1228]: does null_here distinguish between nan/pd.NA/...?

**Comment by lithomas1** [comment]
> OK, looking at this now. Are the asv results from feb 6 still accurate-ish?

Benchmarks still show roughly the same improvement.

File: pandas/_libs/lib.pyx (hunk 21 above)
Comment (lithomas1) [reply pandas/_libs/lib.pyx:1238]: Nope, it shouldn't matter anyways.

The final inferred dtype will determine the type of the na-values.

**Comment by lithomas1** [comment]
@jbrockmendel 
Any chance you can have a look at this soon?

I'm trying to get this in before your deprecation PR.

**Comment by WillAyd** [comment]
How important is this performance boost in the grand scheme of things? I still have some pretty big concerns about the design pattern, loss of type safety and future maintenance burden of this, so I'm not sold on this being merge-able as is

**Comment by jbrockmendel** [comment]
I'll take another look, definitely ping me again (ATM on slack is most likely to not fall through the cracks) if i dont get to it by the weekend.

My knee-jerk reaction mirrors Wills last comment.

**Comment by lithomas1** [comment]
> How important is this performance boost in the grand scheme of things? I still have some pretty big concerns about the design pattern, loss of type safety and future maintenance burden of this, so I'm not sold on this being merge-able as is

The main thing it should be boosting is the speed of ``to_numeric``. If this was ported to ``maybe_convert_objects`` as well, I think it would also boost the perf of constructors and ``infer_objects`` I guess.

The boost isn't really big, but I think the (on average) 20% speed increase and 20% memory usage decrease is not bad.

Happy to let this sit for a while, though, if you're close to making a proposal on changing _libs to Rust(which would guarantee the safetyness of this I guess).

**Comment by jbrockmendel** [comment]
Looking at this again I'm still really uncomfortable with it.  I think a two-pass approach where the first pass does the inference and the second pass does the casting would be way simpler and get all the memory benefits.  How much would that cost perf-wise?

**Comment by mroeschke** [comment]
Seems like there's lukewarm reception to this so far and has stalled. I'll mothball for now but feel free to reopen if interested in a different approach
